model/src/cfd.rs

```diff
@@ -894,10 +903,74 @@ impl Cfd {
         Ok(CfdEvent::new(self.id, EventKind::RolloverStarted))
     }
 
+    pub fn start_rollover_taker(&self) -> Result<CfdEvent> {
+        if self.during_rollover {
+            bail!("The CFD is already being rolled over")
+        };
+
+        self.can_rollover()?;
+
+        let event = CfdEvent::new(self.id, EventKind::RolloverStarted);
+
+        Ok(event)
+    }
+
+    pub fn start_rollover_maker(
+        &self,
+        from_tx_id_proposed: Txid,
+    ) -> Result<(CfdEvent, BitMexPriceEventId, CompleteFee)> {
+        if self.during_rollover {
+            bail!("The CFD is already being rolled over")
+        };
+
+        self.can_rollover()?;
+
+        let dlc = self
+            .dlc
+            .as_ref()
+            .context("No DLC available when starting a rollover")?;
+
+        let current_commit_txid = dlc.commit.0.txid();
+
+        let (from_event_id, from_complete_fee) = if current_commit_txid == from_tx_id_proposed {
+            (dlc.settlement_event_id, self.fee_account.settle())
+        } else {
+            let (from_event_id, from_complete_fee) = match dlc
+                .revoked_commit
+                .iter()
+                .find(|revoke_commit| revoke_commit.txid == from_tx_id_proposed)
+            {
+                Some(revoke_commit) => {
+                    let from_event_id = revoke_commit.settlement_event_id.ok_or_else(|| anyhow!(
+                    "Proposed commit-txid {} not eligible for rollover because no event-id attached", from_tx_id_proposed
+                    ))?;
+                    let from_complete_fee = revoke_commit.complete_fee.ok_or_else(|| anyhow!(
+                        "Proposed commit-txid {} not eligible for rollover because no complete_fee attached", from_tx_id_proposed
+                    ))?;
+
+                    (from_event_id, from_complete_fee)
+                }
+                None => bail!(
+                    "Unknown commit-txid {} proposed by taker",
+                    from_tx_id_proposed
+                ),
+            };
+
+            tracing::info!(order_id=%self.id, commit_txid=%from_tx_id_proposed, %from_event_id, "Starting rollover from previous commit-txid");
+
+            (from_event_id, from_complete_fee)
+        };
+
+        let event = CfdEvent::new(self.id, EventKind::RolloverStarted);
+
+        Ok((event, from_event_id, from_complete_fee))
+    }
+
     pub fn accept_rollover_proposal(
         self,
         tx_fee_rate: TxFeeRate,
         funding_rate: FundingRate,
+        from_params: Option<(BitMexPriceEventId, CompleteFee)>,
```

Comment:
Note that all these are only `Option`s because of backwards compatibility for the legacy protocol versions. 

I would hope to clean that up once we are a few versions ahead. If you feel strong about it feel free to do a small refactoring round as a follow up!